vendor/blockchain_contracts/src/ethereum/rfc003/erc20_htlc.rs

```diff
@@ -0,0 +1,219 @@
+        let amount = <[u8; 32]>::from(self.amount);
+
+        let mut data = [0u8; 4 + 32 + 32];
+        data[..4].copy_from_slice(transfer_fn_abi);
+        data[16..36].copy_from_slice(&htlc_contract_address);
+        data[36..68].copy_from_slice(&amount);
+
+        Bytes::from(data.to_vec())
+    }
+
+    pub fn token_contract_address(&self) -> Address {
+        self.token_contract_address
+    }
+
+    pub fn deployment_gas_limit(&self) -> U256 {
+        let bytes: Bytes = self.compile_to_hex().into();
+        let n_bytes = bytes.0.len();
+        let gas_per_byte = 200;
+
+        U256::from(100_000 + n_bytes * gas_per_byte)
+    }
+
+    pub fn tx_gas_limit() -> U256 {
+        U256::from(100_000)
+    }
+
+    pub fn fund_tx_gas_limit() -> U256 {
+        U256::from(100_000)
+    }
+
+    pub fn compile_template_to_hex() -> String {
+        let code_length = Self::CONTRACT_CODE_TEMPLATE.len() / 2; // In hex, each byte is two chars
+
+        let code_length_as_hex = format!("{:0>4x}", code_length);
+
+        let header_length = Self::DEPLOY_HEADER_TEMPLATE.len() / 2;
+        let header_length_as_hex = format!("{:0>4x}", header_length);
+
+        let deploy_header = Self::DEPLOY_HEADER_TEMPLATE
+            .to_string()
+            .replace(
+                Self::CONTRACT_START_POSITION_PLACEHOLDER,
+                &header_length_as_hex,
+            )
+            .replace(Self::CONTRACT_LENGTH_PLACEHOLDER, &code_length_as_hex);
+
+        deploy_header + Self::CONTRACT_CODE_TEMPLATE
+    }
+
+    fn get_offset(name: &str, placeholder: &str) -> Offset {
+        let contract = Self::compile_template_to_hex();
+
+        let re_match = Regex::new(placeholder)
+            .expect("Could not create regex")
+            .find(contract.as_str())
+            .expect("Could not find placeholder in hex code");
+        Offset::new(
+            name.into(),
+            re_match.start() / 2,
+            re_match.end() / 2,
+            placeholder.len() / 2,
+        )
+    }
+
+    pub fn get_all_offsets() -> Vec<Offset> {
+        let refund_timestamp =
+            Self::get_offset("Refund timestamp", Self::REFUND_TIMESTAMP_PLACEHOLDER);
+        let redeem_address = Self::get_offset("Redeem Address\t", Self::REDEEM_ADDRESS_PLACEHOLDER);
+        let refund_address = Self::get_offset("Refund Address\t", Self::REFUND_ADDRESS_PLACEHOLDER);
+        let secret_hash = Self::get_offset("Secret Hash\t\t", Self::SECRET_HASH_PLACEHOLDER);
+        let amount = Self::get_offset("Amount\t\t\t", Self::AMOUNT_PLACEHOLDER);
+        let token_contract_address =
+            Self::get_offset("Token Contract\t", Self::TOKEN_CONTRACT_ADDRESS_PLACEHOLDER);
+
+        vec![
+            secret_hash,
+            refund_timestamp,
+            redeem_address,
+            refund_address,
+            amount,
+            token_contract_address,
+        ]
+    }
+}
+
+impl Htlc for Erc20Htlc {
+    fn compile_to_hex(&self) -> ByteCode {
```

Comment:
:thinking: 

As soon as this code moves to a different repository, this function should only be used by the tests. From there on, the `comit_node` crate will have to construct the HTLCs by making use of the information from the RFC, e.g. a template + offsets.

Should we therefore remove this functionality right away and test the contracts in the way other people (including us in the `comit_node` crate) will have to construct the HTLCs?

Reply:
I see what you mean. This functionality is still needed in any case to calculate the offsets. this new crate has 2 purposes:
1. Calculate offsets/hex based on assembly code
2. Test contracts

The issue you are raising is that 2. is done with contracted generated from assembly, instead of contracts generated via the offsetting.
This means that in call case we need to keep `compile_to_hex` to do 1.

However, we may want to do 2. based on offsetting as it is the way the contract are constructed in comit_node.
If we were to do 2. with offsets in the `blockchain_contracts` crate then we are duplicating functionality (generation of contract by offset) in `comit_node` and `blockchain_contracts`.

I believe that the following is fine:
`blockchain_contracts`:
- generates offset from assembly
- test contracts built from assembly
`comit_node`:
- build contracts from offset

I am happy to consider the following addition:
`comit_node`:
- tests contracts built from offset

Meaning (some) tests "go back" to `comit_node`.

Let me know what you think

Reply:
Ok agreed, https://github.com/comit-network/comit-rs/issues/896 created

Reply:
❌ This trait shouldn't be implemented here (it should not be imported anywhere in this crate).

When the time comes we simply re-implement this in `comit-node` to compile from the offsets.

Reply:
already done